airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/internals/DebeziumShutdownProcedure.java

```diff
@@ -48,29 +49,35 @@ public DebeziumShutdownProcedure(final LinkedBlockingQueue<T> sourceQueue,
 
   private Runnable transfer() {
     return () -> {
-      while (!sourceQueue.isEmpty() || !MoreBooleans.isTruthy(publisherStatusSupplier.get())) {
+      while (!hasEngineShutDown()) {
         try {
-          T event = sourceQueue.poll(10, TimeUnit.SECONDS);
-          if (event != null) {
-            targetQueue.put(event);
-          }
+          targetQueue.put(sourceQueue.poll(10, TimeUnit.SECONDS));
         } catch (final InterruptedException e) {
           Thread.currentThread().interrupt();
           throw new RuntimeException(e);
         }
       }
+      // At this point, the engine has shut down, and therefore nothing will be writing to the queue any
+      // more. Empty it.
+      sourceQueue.drainTo(targetQueue);
     };
   }
 
+  private boolean hasEngineShutDown() {
+    return MoreBooleans.isTruthy(publisherStatusSupplier.get());
```

Comment:
This method is to check whether the debezium publisher is still running, it's deliberately not related to the queue contents.

Reply:
But logically, is checking that `!sourceQueue.isEmpty()` no longer needed?

Reply:
I don't think so. We only did that in order to decide whether to `poll`. In my commit I only `poll` when the engine is still running. After that, I don't need to, I can just `sourceQueue.drainTo(...)` or whatever because I'll always acquire the lock.

Reply:
See https://github.com/airbytehq/airbyte/pull/31924#discussion_r1374912521